src/main/scala/com/fulcrumgenomics/bam/ZipperBams.scala

```diff
@@ -0,0 +1,191 @@
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.bam.ZipperBams.{buildOutputHeader, checkSort, templateIterator}
+import com.fulcrumgenomics.bam.api.{SamOrder, SamRecord, SamSource, SamWriter}
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.commons.async.AsyncIterator
+import com.fulcrumgenomics.commons.collection.BetterBufferedIterator
+import com.fulcrumgenomics.commons.util.LazyLogging
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.util.Io
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools.{SAMFileHeader, SAMSequenceDictionary}
+import htsjdk.variant.utils.SAMSequenceDictionaryExtractor
+
+/*
+Feature List
+
+    Trim back reads that are FR pairs that extend past each other's starts
+ */
+
+object ZipperBams extends LazyLogging {
+  /** Class to hold info about all the extended tags/attrs we want to manipulate. */
+  case class TagInfo(remove: IndexedSeq[String], reverse: IndexedSeq[String], revcomp: IndexedSeq[String]) {
+    val setToReverse: java.util.Set[String] = remove.iterator.toJavaSet
+    val setToRevcomp: java.util.Set[String] = revcomp.iterator.toJavaSet
+
+    val hasRevsOrRevcomps: Boolean = reverse.nonEmpty || revcomp.nonEmpty
+  }
+
+  /** Checks the source's header declares it is queryname sorted or grouped and logs a warning if not. */
+  def checkSort(source: SamSource, path: FilePath, name: String): Unit = {
+    if (source.header.getSortOrder != SortOrder.queryname && source.header.getGroupOrder != GroupOrder.query) {
+      logger.warning(s"${name} file ${path} does not appear to be queryname sorted or grouped.")
+    }
+  }
+
+  def buildOutputHeader(unmapped: SAMFileHeader,
+                        mapped: SAMFileHeader,
+                        sort: Option[SamOrder],
+                        dict: SAMSequenceDictionary): SAMFileHeader = {
+    // Build a new header using the given sequence dictionary
+    val header = new SAMFileHeader(dict)
+
+    // Copy over comments, RGs and PGs, letting mapped override unmapped if there are conflicts
+    Iterator(unmapped, mapped).foreach { old =>
+      old.getComments.iterator().foreach(header.addComment)
+      old.getReadGroups.iterator().foreach(header.addReadGroup)
+      old.getProgramRecords.iterator().foreach(header.addProgramRecord)
+    }
+
+    // Set the sort and group order
+    header.setSortOrder(unmapped.getSortOrder)
+    header.setGroupOrder(unmapped.getGroupOrder)
+    sort.foreach(s => s.applyTo(header))
+
+    header
+  }
+
+  /** Constructs a template iterator that assumes the input is sorted or grouped, and turns it into an async iter. */
+  def templateIterator(source: SamSource, bufferSize: Int): BetterBufferedIterator[Template] = {
+    val recs = source.iterator.bufferBetter
+    val templates = new Iterator[Template] {
+      override def hasNext: Boolean = recs.hasNext
+      override def next(): Template = Template(recs)
```

Comment:
So I think this will catch the case where the aligner produces multiple primaries for a given read end.  Do we want to have a better error message than what's in `Template`?

Reply:
Yeah, probably.